### PR TITLE
raspberry-pi."4": update poe hat overlay to work with newer kernel

### LINKED ISSUE #626: Applying overlay rpi-poe-plus-overlay fails on current nixpkgs master with can't find symbol 'i2c0if'
After an update of `nixpkgs` I fail to build my raspberry Pi 4 config.

The nixpkgs revision is d4652d33d6af0caf80c9976af34db03d5c2c8fa7
The nixos-hardware revision is 51559e691f1493a26f94f1df1aaf516bb507e78b

The relevant config is:

```
hardware =  raspberry-pi."4".poe-plus-hat.enable = true;
```

and 

```
hardware = {
        enableRedistributableFirmware = true;
        deviceTree = {
          enable = true;
};
```

(They come from different files and get merged later)

The error message I am getting is:

```
./broadcom -> /nix/store/fkj28s2k5bwj9z48hgsfbri6sfslnqj3-device-tree-overlays/./broadcom
'./broadcom/bcm2711-rpi-cm4-io.dtb' -> '/nix/store/fkj28s2k5bwj9z48hgsfbri6sfslnqj3-device-tree-overlays/./broadcom/bcm2711-rpi-cm4-io.dtb'
'./broadcom/bcm2711-rpi-400.dtb' -> '/nix/store/fkj28s2k5bwj9z48hgsfbri6sfslnqj3-device-tree-overlays/./broadcom/bcm2711-rpi-400.dtb'
'./broadcom/bcm2711-rpi-cm4.dtb' -> '/nix/store/fkj28s2k5bwj9z48hgsfbri6sfslnqj3-device-tree-overlays/./broadcom/bcm2711-rpi-cm4.dtb'
'./broadcom/bcm2711-rpi-cm4s.dtb' -> '/nix/store/fkj28s2k5bwj9z48hgsfbri6sfslnqj3-device-tree-overlays/./broadcom/bcm2711-rpi-cm4s.dtb'
'./broadcom/bcm2711-rpi-4-b.dtb' -> '/nix/store/fkj28s2k5bwj9z48hgsfbri6sfslnqj3-device-tree-overlays/./broadcom/bcm2711-rpi-4-b.dtb'
Applying overlay rpi-poe-plus-overlay to bcm2711-rpi-cm4-io.dtb... DTOVERLAY[error]: can't find symbol 'i2c0if'
```

If I disable the `poe` overlay, the build succeeds. 

I suspect it has something to do with an update of the kernel? I'm not really sure how I can debug this failure. A quick google and discourse search didn't yield any current reports.

### PR DESCRIPTION
**Pull request by gador**
fixes #626

###### Description of changes

This updates the `poe` dts overlays to work with the newer kernel (6.1) on `nixpkgs` master.

~~I am unsure on how to keep `nixos-hardware` and `nixpkgs` in sync here. I suspect this update will break existing configs which aren't updated to `nixpkgs` master..~~

~~I also haven't tested the updated config on my raspberry pi 4, yet. But the config builds and _should_ work~~

EDIT:

With the assertion this PR should at least warn and not build with an incompatible kernel. 

I have also just tested this PR on my RP4b and the fan of the POE hat works without issues.

###### Things done

- [x] Tested the changes in your own NixOS Configuration
- [x] Tested the changes end-to-end by using your fork of `nixos-hardware` and
      importing it via `<nixos-hardware>` or Flake input

### CODE REVIEW

File: raspberry-pi/4/poe-hat.nix [hunk 1]
```diff
@@ -16,7 +16,7 @@ in {
 
     hardware.deviceTree = {
       overlays = [
-        # Equivalent to: https://github.com/raspberrypi/linux/blob/rpi-5.15.y/arch/arm/boot/dts/overlays/rpi-poe-overlay.dts
+        # Equivalent to: https://github.com/raspberrypi/linux/blob/rpi-6.1.y/arch/arm/boot/dts/overlays/rpi-poe-overlay.dts
```
Comment (Mic92) [inline raspberry-pi/4/poe-hat.nix:19]: Can you set an assertion like here: https://github.com/NixOS/nixpkgs/blob/8a05bed379935ad7e50b338edd09f7f13c996c6a/nixos/modules/services/logging/journalbeat.nix#L67
to make sure the linux is >= 6.1 (i.e. `(lib.versionAtLeast config.boot.kernelPackages.kernel "6.1")`)

Comment (Mic92) [reply raspberry-pi/4/poe-hat.nix:19]: Than people on stable can upgrade.

**Comment by gador** [comment]
The test fail due to the new assertion. (The kernel from the test is older than 6.1)

Also, I had to remove the compute module from the POE overlay, since it doesn't have the required symbols (which makes sense; The POE hat can't be stacked upon a compute module)

**Comment by Mic92** [comment]
bors merge

**Comment by mirkolenz** [comment]
Thank you for keeping the PoE-Hat overlays up-to-date!

Sadly, my NixOS configuration on a Raspberry Pi 4 broke today due to these changes (it now requires a kernel version newer than 6.1). I am targeting the branch nixos-22.11 for my nixpkgs flake input, the Raspi Kernel with v6.1 is only available in unstable currently (and will most likely be included in 23.05). How should I handle this situation in the next weeks? Should I force the use of the current unstable kernel (the default is here: https://github.com/NixOS/nixos-hardware/blob/master/raspberry-pi/4/default.nix#L20)?

Initially I tried to find a "release-22.11" branch in this repo (this is also how Home Manager handles such incompatible changes), but it seems that there is no such branch.

Thank you very much in advance!

**Comment by mirkolenz** [comment]
No worries, I really appreciate your effort in making the overlays compatible with the newer kernel. I will try to use the unstable kernel for my Raspberry Pi. Maybe it would be a good idea to set up separate release branches for `nixos-hardware` (like `nixpkgs` and `home-manager` do it) in the future?

**Comment by purcell** [comment]
I have `hardware.raspberry-pi."4".fkms-3d.enable = true;` and found that both before and after this PR, the deviceTree failed to build with recent `unstable` Nixpkgs. The error was the generic "FDT_..." one which wasn't very helpful. I tried changing the default device tree filter too, as you have done in this PR (adding the `-4*` in the filter pattern`, but the resulting wasn't bootable, for reasons I couldn't fathom. Debugging it all has been a bit beyond me today, and I resorted to painstakingly restoring an older working boot configuration. Any pointers welcome!

**Comment by mirkolenz** [comment]
Just FYI, I tried using the `nixos-unstable` branch for the kernel packages, but my config fails to build:

```txt
error: builder for '/nix/store/1vzn6pd984vwc2ylsl6sv2xqqjy8n38s-device-tree-overlays.drv' failed with exit code 1;
       last 6 log lines:
       > ./broadcom -> /nix/store/9zrcl7cdhfc30v1dp8dk8csfa9db0k9b-device-tree-overlays/./broadcom
       > './broadcom/bcm2711-rpi-400.dtb' -> '/nix/store/9zrcl7cdhfc30v1dp8dk8csfa9db0k9b-device-tree-overlays/./broadcom/bcm2711-rpi-400.dtb'
       > './broadcom/bcm2711-rpi-4-b.dtb' -> '/nix/store/9zrcl7cdhfc30v1dp8dk8csfa9db0k9b-device-tree-overlays/./broadcom/bcm2711-rpi-4-b.dtb'
       > Applying overlay rpi-poe-plus-overlay to bcm2711-rpi-400.dtb... ok
       > Skipping overlay rpi4-cpu-revision: incompatible with bcm2711-rpi-400.dtb
       > Applying overlay audio-on-overlay to bcm2711-rpi-400.dtb... DTOVERLAY[error]: can't find symbol 'audio'
       For full logs, run 'nix log /nix/store/1vzn6pd984vwc2ylsl6sv2xqqjy8n38s-device-tree-overlays.drv'.
error: 1 dependencies of derivation '/nix/store/xb04lg1j7ws4kkbf1l3l132k3lc7190c-nixos-system-raspi-22.11.20230515.55af203.drv' failed to build
```

So it seems that there are some incompatibilities.

**Comment by gador** [comment]
> Applying overlay audio-on-overlay

This PR didn't touch this overlay. Maybe it needs an update to be compatible with the newer kernel as well?

**Comment by gador** [comment]
> I have `hardware.raspberry-pi."4".fkms-3d.enable = true;` and found that both before and after this PR, the deviceTree failed to build with recent `unstable` Nixpkgs. The error was the generic "FDT_..." one which wasn't very helpful. I tried changing the default device tree filter too, as you have done in this PR (adding the `-4*` in the filter pattern`, but the resulting wasn't bootable, for reasons I couldn't fathom. Debugging it all has been a bit beyond me today, and I resorted to painstakingly restoring an older working boot configuration. Any pointers welcome!

This has been an issue for me, too (independently of this PR).
I just disabled the fkms overlay..

You could try bisecting nixpkgs or rebuild the config to different nixpkgs unstable revisions to get an idea of what changed. This will be a lot of work, though

**Comment by purcell** [comment]
> I suggest creating a new issue as it seems unrelated to this PR

Yeah, I hesitated to do that because I didn't have the build failure logs anymore, but that's indeed the best plan: #631

**Comment by purcell** [comment]
PS. My Pi4 is headless, but I *think* I need the fkms overlay for accelerated video encoding/decoding... perhaps I'm completely mistaken though?

**Comment by Nebucatnetzer** [comment]
> My Pi4 is headless, but I think I need the fkms overlay for accelerated video encoding/decoding

This is my understanding as well.

**Comment by calbrecht** [comment]
Uhm, excuse my ignorance, with `hardware.raspberry-pi."4".apply-overlays-dtmerge.enable = true` the error changes to indicate the symbol '**fb**' wasn't found. When modifying `hardware.deviceTree.overlays` with removing the first fragment from `rpi4-vc4-fkms-v3d-overlay` it cannot find the symbol '**firmwarekms**'. When also removing the second fragment, it builds the device tree just fine. However, i really don't understand the consequences there, just thought some knowledgeable person could take it further from here.

**Pull request merged**